src/main/kotlin/org/rust/ide/annotator/RsSyntaxErrorsAnnotator.kt

```diff
@@ -277,17 +277,39 @@ private fun checkTypeParameterList(holder: AnnotationHolder, element: RsTypePara
                     .range(it).create()
             }
     }
-    val lifetimeParams = element.lifetimeParameterList
-    if (lifetimeParams.isEmpty()) return
-    val startOfTypeParams = element.typeParameterList.firstOrNull()?.textOffset ?: return
-    for (e in lifetimeParams) {
-        if (e.textOffset > startOfTypeParams) {
-            holder.newAnnotation(HighlightSeverity.ERROR, "Lifetime parameters must be declared prior to type parameters")
-                .range(e).create()
+
+    var kind = TypeParameterKind.LIFETIME
+    element.forEachChild { child ->
+        val newKind = TypeParameterKind.forType(child) ?: return@forEachChild
+        if (newKind.canStandAfter(kind)) {
+            kind = newKind
+        } else {
+            val newStateName = newKind.parameterNameCapitalized
+
+            holder.newAnnotation(HighlightSeverity.ERROR, "$newStateName must be declared prior to ${kind.parameterName}")
```

Comment:
Perhaps similar content from PR can be extracted into the localization file now to ease the work later.

Reply:
Not sure that we should do it in this PR
I suppose that it should be done in separate PRs with extracting all string literals for a particular subsystem into property files and declaring rules not to add new hardcoded literals in code.
But I still don't fully understand how we will support translations in an actual state